R/weight.R

```diff
@@ -59,3 +64,111 @@ setMethod("weightVariables", "VariableCatalog", function (x) {
         return(c())
     }
 })
+
+#' Generate a weight variable
+#'
+#' This function allows you to generate a weight vector by supplying a set of
+#' categorical variables and the target distribution for each of the variable's categories. Weights are
+#' computed by iteratively 'raking' conditional 'cells' to marginal population targets.
+#' For instance if you wanted to create a weight variable which equally weighted four categories stored
+#' in `ds$var` you would call `ds$weight1 <- makeWeight(ds$var ~ c(25, 25, 25, 25), name = "weight1")`. This would
+#' create a new variable which can then be set as the dataset weight with [weight]. You can also create
+#' the dataset weights in one step with `weight(ds) <- makeWeight(ds$var ~ c(25, 25, 25, 25), name = "weight1")`.
+#'
+#' @param ...
+#' A series of expressions of the form `variable ~ target_weights`. The variable must
+#' be a categorical Crunch variable, and the target weights must be a numeric vector whose
+#' length should be equal to the number of categories contained in the variable, and whose sum is equal to 100 or 1. If
+#' you supply fewer target weights than there are categories `makeWeight` will pad the target weight vector with 0s.
+#'
+#' @param name
+#' The name of the resulting variable
+#' @rdname makeWeight
+#' @return
+#' A crunch [VariableDefinition()] of the weight variable
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' mtcars$cyl <- as.factor(mtcars$cyl)
+#' mtcars$gear <- as.factor(mtcars$gear)
+#' ds <- newDataset(mtcars)
+#' ds$weight <- makeWeight(ds$cyl ~ c(30, 30, 40, 0), ds$gear ~ c(20, 20, 60, 0), name = "weight" )
+#' summary(ds$weight)
+#' weight(ds) <-  makeWeight(ds$var ~ c(25, 25, 25, 25), name = "weight2")
+#' }
+makeWeight <- function(..., name) {
+    all_dots <- list(..., name = name)
+    named_entries <- names(all_dots) != ""
+    out       <- all_dots[named_entries]
+    expr_list <- all_dots[!named_entries]
+
+    #args below must be an unnamed list, so we remove the names from expr_list
+    names(expr_list) <- NULL
+    out$derivation <- list(
+            `function` = "rake",
+            args = lapply(expr_list, generateWeightEntry)
+        )
+    class(out) <- "VariableDefinition"
+    out
+}
+
+#' Generate entry for [makeWeight]
+#'
+#' Utility function to catch formula errors passed to [makeWeight], and to generate
+#' the appropriate entry.
+#'
+#' @param expr
+#' An expression
+#' @return
+#' A list of the variable id and the target weights.
+#'
+#' @keywords internal
+#' @examples
+#' \dontrun{
+#' validateWeightExpression(ds$var ~ c(10, 30, 60))
+#' }
+#'
+generateWeightEntry <- function(expr) {
+
+    formula <- try(as.formula(expr), silent = TRUE)
+    if (is.error(formula)) {
+        halt(dQuote(substitute(expr)), " is not a valid formula. Use the form ds$var ~ c(10, 20, 30)")
```

Comment:
Minor point, but perhaps in the example `ds$var ~ c(10, 20, 30)`, the percentages should add to 100 as is required.